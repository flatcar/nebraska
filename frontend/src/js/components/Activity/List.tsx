import MuiList from '@material-ui/core/List';
import { makeStyles } from '@material-ui/core/styles';
import Typography from '@material-ui/core/Typography';
import React from 'react';
import { Activity } from '../../api/apiDataTypes';
import { makeLocaleTime } from '../../constants/helpers';
import Item from './Item';

const useStyles = makeStyles(theme => ({
  listTitle: {
    fontSize: '1em',
  },
}));

function List(props: {entries?: Activity[]; timestamp: string}) {
  const classes = useStyles();
  const entries = props.entries ? props.entries : [];

  return (
    <React.Fragment>
      <Typography className={classes.listTitle}>
        {makeLocaleTime(props.timestamp, {
          showTime: false,
          dateFormat: {weekday: 'long', month: 'long', day: 'numeric', year: 'numeric'}
        })}
      </Typography>
      <MuiList>
        {entries.map((entry: Activity, i: number) =>
          <Item key={i} entry={entry} />
        )}
      </MuiList>
    </React.Fragment>
  );
}

export default List;
