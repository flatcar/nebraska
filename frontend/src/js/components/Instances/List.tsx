import { makeStyles, Theme } from '@material-ui/core';
import Box from '@material-ui/core/Box';
import Button from '@material-ui/core/Button';
import FormControl from '@material-ui/core/FormControl';
import Grid from '@material-ui/core/Grid';
import Input from '@material-ui/core/Input';
import InputLabel from '@material-ui/core/InputLabel';
import MenuItem from '@material-ui/core/MenuItem';
import Paper from '@material-ui/core/Paper';
import Select from '@material-ui/core/Select';
import TablePagination from '@material-ui/core/TablePagination';
import { useTheme } from '@material-ui/styles';
import PropTypes from 'prop-types';
import React, { ChangeEvent } from 'react';
import { useHistory, useLocation } from 'react-router-dom';
import API from '../../api/API';
import { Application, Group } from '../../api/apiDataTypes';
import { getInstanceStatus, useGroupVersionBreakdown } from '../../constants/helpers';
import Empty from '../Common/EmptyContent';
import ListHeader from '../Common/ListHeader';
import Loader from '../Common/Loader';
import TimeIntervalLinks from '../Common/TimeIntervalLinks';
import { InstanceCountLabel } from './Common';
import makeStatusDefs from './StatusDefs';
import Table from './Table';

const useStyles = makeStyles(theme => ({
  root: {
    backgroundColor: theme.palette.lightSilverShade
  }
}));

interface InstanceFilterProps {
  versions: any[];
  onFiltersChanged: (newFilters: any) => void;
  filter: {
    [key: string]: any;
  };
  disabled?: boolean;
}

function InstanceFilter(props: InstanceFilterProps) {
  const statusDefs = makeStatusDefs(useTheme());
  const {onFiltersChanged, versions} = props;

  function changeFilter(filterName: string, filterValue: string) {
    if (filterValue === props.filter[filterName]) {
      return;
    }

    const filter = {...props.filter};
    filter[filterName] = filterValue;

    onFiltersChanged(filter);
  }

  return (
    <Box pr={2}>
      <Grid container spacing={2} justify="flex-end">
        <Grid item xs={5}>
          <FormControl
            fullWidth
            disabled={props.disabled}
          >
            <InputLabel htmlFor="select-status" shrink>Filter Status</InputLabel>
            <Select
              onChange={(event: any) => changeFilter('status', event.target.value) }
              input={<Input id="select-status" />}
              renderValue={(selected: any) =>
                selected ? statusDefs[selected].label : 'Show All'
              }
              value={props.filter.status}
              displayEmpty
            >
              <MenuItem key="" value="">Show All</MenuItem>
              {
                Object.keys(statusDefs).map(statusType => {
                  const label = statusDefs[statusType].label;
                  return <MenuItem key={statusType} value={statusType}>{label}</MenuItem>;
                })
              }
            </Select>
          </FormControl>
        </Grid>
        <Grid item xs={5}>
          <FormControl
            fullWidth
            disabled={props.disabled}
          >
            <InputLabel htmlFor="select-versions" shrink>Filter Version</InputLabel>
            <Select
              onChange={(event: ChangeEvent<{name?: string | undefined; value: any}>) => changeFilter('version', event.target.value) }
              input={<Input id="select-versions" />}
              renderValue={(selected: any) =>
                selected ? selected : 'Show All'
              }
              value={props.filter.version}
              displayEmpty
            >
              <MenuItem key="" value="">Show All</MenuItem>
              {
                (versions || []).map(({version}) => {
                  return <MenuItem key={version} value={version}>{version}</MenuItem>;
                })
              }
            </Select>
          </FormControl>
        </Grid>
      </Grid>
    </Box>
  );
}

function ListView(props: {application: Application; group: Group}) {
  const classes = useStyles();
  const theme = useTheme();
  const statusDefs = makeStatusDefs(useTheme());
  const {application, group} = props;
  const versionBreakdown = useGroupVersionBreakdown(group);
  /*TODO: use the URL as the single source of truth and remove states */
  const [page, setPage] = React.useState(0);
  const [rowsPerPage, setRowsPerPage] = React.useState(10);
  const [filters, setFilters] = React.useState<{[key: string]: any}>({status: '', version: ''});
  const [instancesObj, setInstancesObj] = React.useState({instances: [], total: -1});
  const [instanceFetchLoading, setInstanceFetchLoading] = React.useState(false);
  const [totalInstances, setTotalInstances] = React.useState(-1);
  const location = useLocation();
  const history = useHistory();

  function getDuration() {
    return (new URLSearchParams(location.search)).get('period') || '1d';
  }

  function fetchFiltersFromURL(callback: (...args: any) => void) {
    let status = '';
    const queryParams = new URLSearchParams(location.search);
    if (queryParams.has('status')) {
      const statusValue = queryParams.get('status');
      if (statusValue !== 'ShowAll') {
        for (const key in statusDefs) {
          if (statusDefs[key].label === statusValue) {
            status = key;
            break;
          }
        }
      }
    }
    const version = queryParams.get('version') || '';
    const pageQueryParam = (parseInt(queryParams.get('page') as string) || 1) - 1;
    const perPage = parseInt(queryParams.get('perPage') as string) || 10;
    const duration = getDuration();

    callback(status, version, pageQueryParam, perPage, duration);
  }

  function addQuery(queryObj: {[key: string]: any}) {
    const pathname = location.pathname;
    const searchParams = new URLSearchParams(location.search);
    for (const key in queryObj) {
      const value = queryObj[key];
      if (value) {
        searchParams.set(key, value);
      } else {
        searchParams.delete(key);
      }
    }

    history.push({
      pathname: pathname,
      search: searchParams.toString()
    });
  };

  function fetchInstances(filters: {[key: string]: any}, page: number,
                          perPage: number, duration: string) {
    setInstanceFetchLoading(true);
    const fetchFilters = {...filters};
    if (filters.status === '') {
      fetchFilters.status = '0';
    } else {
      const statusDefinition = statusDefs[fetchFilters.status];
      fetchFilters.status = statusDefinition
        .queryValue;
    }
    API.getInstances(application.id, group.id,
      {
        ...fetchFilters,
        page: page + 1,
        perpage: perPage,
        duration
      }).then((result) => {
      setInstanceFetchLoading(false);
      // Since we have retrieved the instances without a filter (i.e. all instances)
      // we update the total.
      if (!fetchFilters.status && !fetchFilters.version) {
        setTotalInstances(result.total);
      }
      if (result.instances) {
        const massagedInstances = result.instances.map((instance: any) => {
          instance.statusInfo = getInstanceStatus(instance.application.status);
          return instance;
        });
        setInstancesObj({instances: massagedInstances, total: result.total});
      } else {
        setInstancesObj({instances: [], total: result.total});
      }
    })
      .catch(() => {
        setInstanceFetchLoading(false);
      });
  }

  function handleChangePage(event: React.MouseEvent<HTMLButtonElement, MouseEvent> | null,
                            newPage: number) {
    addQuery({ page: newPage + 1 });
  }

  function handleChangeRowsPerPage(event: React.ChangeEvent<{value: string}>) {
    addQuery({ page: 1, perPage: +event.target.value });
  }

  function onFiltersChanged(newFilters: {[key: string]: any}) {
    applyFilters(newFilters);
  }

  function applyFilters(_filters = {}) {
    const newFilters: {[key: string]: any} = Object.keys(_filters).length !== 0 ?
      _filters : {status: '', version: ''};
    const statusQueryParam = newFilters.status ? statusDefs[newFilters.status].label : '';
    addQuery({ status: statusQueryParam, version: newFilters.version });
    setFilters(newFilters);
  }

  function resetFilters() {
    applyFilters();
  }

  React.useEffect(() => {
    fetchFiltersFromURL((status: string, version: string,
                         pageParam: number, perPageParam: number, duration: string) => {
      setFilters({status, version});
      setPage(pageParam);
      setRowsPerPage(perPageParam);
      fetchInstances({status, version}, pageParam, perPageParam, duration);
    });
  }, [location]);

  React.useEffect(() => {
    // We only want to run it once ATM.
    if (totalInstances > 0) {
      return;
    }

    // We use this function without any filter to get the total number of instances
    // in the group.
    const queryParams = new URLSearchParams(window.location.search);
    const duration = queryParams.get('period');
    API.getInstancesCount(application.id, group.id, duration as string)
      .then(result => {
        setTotalInstances(result);
      })
      .catch(err => console.error('Error loading total instances in Instances/List', err));
  },
  [totalInstances]);

  function getInstanceCount() {
    const total = totalInstances > -1 ? totalInstances : 'â€¦';
    const instancesTotal = instancesObj.total > -1 ? instancesObj.total : '...';
    if (!filters.status && !filters.version || instancesTotal === total) {
      return total;
    }
    return `${instancesTotal}/${total}`;
  }

  function isFiltered() {
    return filters.status || filters.version;
  }
  return (
    <>
      <ListHeader
        title="Instance List"
      />
      <Paper>
        <Box padding="1em">
          <Grid
            container
            spacing={1}
          >
            <Grid
              item
              container
              justify="space-between"
              alignItems="stretch"
            >
              <Grid item>
                <Box mb={2}
                  color={(theme as Theme).palette.greyShadeColor}
                  fontSize={30}
                  fontWeight={700}
                >
                  {group.name}
                </Box>
              </Grid>
              <Grid item>
                <TimeIntervalLinks
                  intervalChangeHandler={(duration) => addQuery({period: duration.queryValue})}
                  selectedInterval={getDuration()}
                  appID={application.id}
                  groupID={group.id}
                />
              </Grid>
            </Grid>
            <Box width="100%" borderTop={1} borderColor={'#E0E0E0'} className={classes.root}>
              <Grid
                item
                container
                md={12}
                alignItems="stretch"
                justify="space-between"
              >
                <Grid item md>
                  <Box ml={2}>
                    <InstanceCountLabel
                      countText={getInstanceCount()}
                      instanceListView
                    />
                  </Box>
                </Grid>
                <Grid item md>
                  <Box mt={2}>
                    <InstanceFilter
                      versions={versionBreakdown}
                      onFiltersChanged={onFiltersChanged}
                      filter={filters}
                    />
                  </Box>
                </Grid>
              </Grid>
            </Box>
            {isFiltered() &&
            <Grid item md={12} container justify="center">
              <Grid item>
                <Button
                  variant="outlined"
                  color="secondary"
                  onClick={resetFilters}
                >
                  Reset filters
                </Button>
              </Grid>
            </Grid>
            }
            <Grid item md={12}>
              {!instanceFetchLoading ?
                (instancesObj.instances.length > 0 ?
                  <React.Fragment>
                    <Table
                      channel={group.channel}
                      instances={instancesObj.instances}
                    />
                    <TablePagination
                      rowsPerPageOptions={[10, 25, 50, 100]}
                      component="div"
                      count={instancesObj.total}
                      rowsPerPage={rowsPerPage}
                      page={page}
                      backIconButtonProps={{
                        'aria-label': 'previous page',
                      }}
                      nextIconButtonProps={{
                        'aria-label': 'next page',
                      }}
                      onChangePage={handleChangePage}
                      onChangeRowsPerPage={handleChangeRowsPerPage}
                    />
                  </React.Fragment>
                  :
                  <Empty>No instances.</Empty>
                )
                :
                  <Loader />
              }
            </Grid>
          </Grid>
        </Box>
      </Paper>
    </>
  );
}

ListView.propTypes = {
  application: PropTypes.object.isRequired,
  group: PropTypes.object.isRequired,
};

export default ListView;
