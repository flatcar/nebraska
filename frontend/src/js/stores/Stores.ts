import ActivityStore from './ActivityStore';
import ApplicationsStore from './ApplicationsStore';
import GroupChartsStore from './GroupChartsStore';

export const applicationsStore = new ApplicationsStore;
export const activityStore = new ActivityStore;
export const groupChartStore = new GroupChartsStore();
