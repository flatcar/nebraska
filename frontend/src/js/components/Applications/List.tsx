import {List as MuiList, withStyles} from '@material-ui/core';
import Paper from '@material-ui/core/Paper';
import React from 'react';
import _ from 'underscore';
import { applicationsStore, modalStore } from '../../stores/Stores';
import Empty from '../Common/EmptyContent';
import ListHeader from '../Common/ListHeader';
import Loader from '../Common/Loader';
import ModalButton from '../Common/ModalButton.react';
import EditDialog from './EditDialog';
import Item from './Item.react';

const styles = theme => ({
  root: {
    '& > hr:first-child': {
      display: 'none'
    }
  }
});

class List extends React.Component {

  constructor(props) {
    super(props);
    this.onChange = this.onChange.bind(this);
    this.searchUpdated = this.searchUpdated.bind(this);
    this.openUpdateAppModal = this.openUpdateAppModal.bind(this);
    this.closeUpdateAppModal = this.closeUpdateAppModal.bind(this);

    this.state = {
      applications: applicationsStore.getCachedApplications(),
      searchTerm: '',
      updateAppModalVisible: false,
      updateAppIDModal: null
    };
  }

  closeUpdateAppModal() {
    this.setState({updateAppModalVisible: false});
  }

  openUpdateAppModal(appID) {
    this.setState({updateAppModalVisible: true, updateAppIDModal: appID});
  }

  componentDidMount() {
    applicationsStore.addChangeListener(this.onChange);
  }

  componentWillUnmount() {
    applicationsStore.removeChangeListener(this.onChange);
  }

  onChange() {
    this.setState({
      applications: applicationsStore.getCachedApplications()
    });
  }

  searchUpdated(event) {
    const {name, value} = event.currentTarget;
    this.setState({searchTerm: value.toLowerCase()});
  }

  render() {
    let applications = this.state.applications;
    let entries = '';
    const {classes} = this.props;

    if (this.state.searchTerm) {
      applications = applications.filter(app => app.name
        .toLowerCase()
        .includes(this.state.searchTerm));
    }

    if (_.isNull(applications)) {
      entries = <Loader />;
    } else {
      if (_.isEmpty(applications)) {
        if (this.state.searchTerm) {
          entries = <Empty>No results found.</Empty>;
        } else {
          entries = <Empty>Ops, it looks like you have not created any
            application yet..<br/><br/> Now is a great time to create your
            first one, just click on the plus symbol above.</Empty>;
        }
      } else {
        entries = _.map(applications, (application, i) => {
          return (
            <Item key={application.id}
              application={application}
              handleUpdateApplication={this.openUpdateAppModal}
            />);
        });
      }
    }

    const appToUpdate = applications &&
      this.state.updateAppIDModal ?
      _.findWhere(applications, {id: this.state.updateAppIDModal}) : null;
    return (
      <>
        <ListHeader
          title="Applications"
          actions={[
            <ModalButton
              modalToOpen="AddApplicationModal"
              data={{applications: applications}}
            />
          ]}
        />
        <Paper>
          <MuiList className={classes.root}>
            {entries}
          </MuiList>
          {appToUpdate &&
          <EditDialog
            data={appToUpdate}
            show={this.state.updateAppModalVisible}
            onHide={this.closeUpdateAppModal}
          />
          }
        </Paper>
      </>
    );
  }

}

export default withStyles(styles)(List);
