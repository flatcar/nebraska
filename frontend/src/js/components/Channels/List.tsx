import { Box, makeStyles } from '@material-ui/core';
import Grid from '@material-ui/core/Grid';
import MuiList from '@material-ui/core/List';
import ListSubheader from '@material-ui/core/ListSubheader';
import Typography from '@material-ui/core/Typography';
import React from 'react';
import _ from 'underscore';
import API from '../../api/API';
import { Application, Channel, Package } from '../../api/apiDataTypes';
import { ARCHES } from '../../constants/helpers';
import { applicationsStore } from '../../stores/Stores';
import Loader from '../Common/Loader';
import ModalButton from '../Common/ModalButton';
import SectionPaper from '../Common/SectionPaper';
import EditDialog from './EditDialog';
import Item from './Item';

const useStyles = makeStyles({
  root: {
    '& > hr:first-child': {
      display: 'none'
    }
  }
});

interface PackageChannelApplication extends Application {
  packages: Package[];
  channels: Channel[];
}

function ChannelList(props: {
  application: PackageChannelApplication;
  onEdit: (channelId: string) => void;
}) {
  const {application, onEdit} = props;
  const classes = useStyles();

  function getChannelsPerArch() {
    const perArch: {
      [key: number]: any[];
    } = {};

    // If application doesn't have any channel return empty object.
    if (application.channels === null) {
      return perArch;
    }

    application.channels.forEach((channel: Channel) => {
      if (!perArch[channel.arch]) {
        perArch[channel.arch] = [];
      }
      perArch[channel.arch].push(channel);
    });

    return perArch;
  }

  return (
    <React.Fragment>
      {Object.entries(getChannelsPerArch()).map(([arch, channels]) =>
        <MuiList
          key={arch}
          subheader={<ListSubheader disableSticky >{ARCHES[parseInt(arch)]}</ListSubheader>}
          dense
          className={classes.root}
        >
          {channels.map(channel =>
            <Item
              key={'channelID_' + channel.id}
              channel={channel}
              packages={application.packages || []}
              showArch={false}
              handleUpdateChannel={onEdit}
            />
          )}
        </MuiList>
      )}
    </React.Fragment>
  );
}

function List(props: {
  appID: string;
}) {
  const { appID } = props;
  const [application, setApplication] =
    React.useState(applicationsStore.getCachedApplication(appID));
  const [packages, setPackages] = React.useState<null | Package[]>(null);
  const [channelToEdit, setChannelToEdit] = React.useState<null | Channel>(null);

  React.useEffect(() => {
    applicationsStore.addChangeListener(onStoreChange);

    // In case the application was not yet cached, we fetch it here
    if (application === null) {
      applicationsStore.getApplication(props.appID);
    }

    // Fetch packages
    if (!packages){
      API.getPackages(props.appID)
        .then((result) => {
          if (_.isNull(result)) {
            setPackages([]);
            return;
          }
          setPackages(result);
        });
    }

    return function cleanup() {
      applicationsStore.removeChangeListener(onStoreChange);
    };
  },
  [appID]);

  function onStoreChange() {
    setApplication(applicationsStore.getCachedApplication(appID));
  }

  function onChannelEditOpen(channelID: string) {
    let channels = [];
    if (application) {
      channels = application.channels ? application.channels : [];
    }

    const channelToUpdate = (!_.isEmpty(channels) && channelID) ?
      _.findWhere(channels, {id: channelID}) : null;

    setChannelToEdit(channelToUpdate);
  }

  function onChannelEditClose() {
    setChannelToEdit(null);
  }

  return (
    <Box mt={2}>
      <Box mb={2}>
        <Grid
          container
          alignItems="center"
          justify="space-between"
        >
          <Grid item>
            <Typography variant="h4">Channels</Typography>
          </Grid>
          <Grid item>
            <ModalButton
              modalToOpen="AddChannelModal"
              data={{
                packages: packages,
                applicationID: appID
              }}
            />
          </Grid>
        </Grid>
      </Box>
      <SectionPaper>
        {!application ?
          <Loader />
          :
          <ChannelList
            application={application}
            onEdit={onChannelEditOpen}
          />
        }
        {channelToEdit &&
        <EditDialog
          data={{packages: packages, applicationID: appID, channel: channelToEdit}}
          show={channelToEdit !== null}
          onHide={onChannelEditClose}
        />
        }
      </SectionPaper>
    </Box>
  );
}

export default List;
