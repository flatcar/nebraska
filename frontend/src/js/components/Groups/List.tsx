import { withStyles } from '@material-ui/core';
import Grid from '@material-ui/core/Grid';
import MuiList from '@material-ui/core/List';
import Paper from '@material-ui/core/Paper';
import PropTypes from 'prop-types';
import React from 'react';
import _ from 'underscore';
import { applicationsStore } from '../../stores/Stores';
import Empty from '../Common/EmptyContent';
import ListHeader from '../Common/ListHeader';
import SearchInput from '../Common/ListSearch';
import Loader from '../Common/Loader';
import ModalButton from '../Common/ModalButton';
import EditDialog from './EditDialog';
import Item from './Item.react';

const styles = theme => ({
  root: {
    '& > hr:first-child': {
      display: 'none'
    }
  }
});

class List extends React.Component {

  constructor(props) {
    super(props);
    this.onChange = this.onChange.bind(this);
    this.searchUpdated = this.searchUpdated.bind(this);
    this.openUpdateGroupModal = this.openUpdateGroupModal.bind(this);
    this.closeUpdateGroupModal = this.closeUpdateGroupModal.bind(this);

    this.state = {
      application: applicationsStore.getCachedApplication(this.props.appID),
      searchTerm: '',
      updateGroupModalVisible: false,
      updateGroupIDModal: null,
      updateAppIDModal: null
    };
  }

  closeUpdateGroupModal() {
    this.setState({updateGroupModalVisible: false});
  }

  openUpdateGroupModal(appID, groupID) {
    this.setState({
      updateGroupModalVisible: true,
      updateGroupIDModal: groupID,
      updateAppIDModal: appID
    });
  }

  componentDidMount() {
    applicationsStore.addChangeListener(this.onChange);
  }

  componentWillUnmount() {
    applicationsStore.removeChangeListener(this.onChange);
  }

  onChange() {
    this.setState({
      application: applicationsStore.getCachedApplication(this.props.appID)
    });
  }

  searchUpdated(event) {
    const {name, value} = event.currentTarget;
    this.setState({searchTerm: value.toLowerCase()});
  }

  render() {
    const application = this.state.application;

    let channels = [];
    let groups = [];
    let packages = [];
    let instances = 0;
    let name = '';
    let entries = '';

    if (application) {
      name = application.name;
      groups = application.groups ? application.groups : [];
      packages = application.packages ? application.packages : [];
      instances = application.instances ? application.instances : [];
      channels = application.channels ? application.channels : [];

      if (this.state.searchTerm) {
        groups = groups.filter(app => app.name.toLowerCase().includes(this.state.searchTerm));
      }

      if (_.isEmpty(groups)) {
        if (this.state.searchTerm) {
          entries = <Empty>No results found.</Empty>;
        } else {
          entries = <Empty>
            There are no groups for this application yet.
            <br/><br/>
            Groups help you control how you want to distribute updates to
            a specific set of instances.
          </Empty>;
        }
      } else {
        entries = _.map(groups, (group, i) => {
          return (
            <Item
              key={'groupID_' + group.id}
              group={group}
              appName={name}
              channels={channels}
              handleUpdateGroup={this.openUpdateGroupModal}
            />);
        });
      }

    } else {
      entries = <Loader />;
    }

    const groupToUpdate = !_.isEmpty(groups) &&
      this.state.updateGroupIDModal ?
      _.findWhere(groups, {id: this.state.updateGroupIDModal}) : null;
    const {classes} = this.props;

    return (
      <>
        <ListHeader
          title="Groups"
          actions={[
            <ModalButton
              icon="plus"
              modalToOpen="AddGroupModal"
              data={{
                channels: channels,
                appID: this.props.appID
              }}
            />
          ]}
        />
        <Paper>
          <MuiList className={classes.root}>
            {entries}
          </MuiList>
          {groupToUpdate &&
          <EditDialog
            data={{group: groupToUpdate, channels: channels}}
            show={this.state.updateGroupModalVisible}
            onHide={this.closeUpdateGroupModal}
          />
          }
        </Paper>
      </>
    );
  }

}

List.propTypes = {
  appID: PropTypes.string.isRequired
};

export default withStyles(styles)(List);
