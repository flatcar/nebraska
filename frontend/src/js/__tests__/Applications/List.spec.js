import {render, waitForDomChange} from '@testing-library/react';
import React from 'react';
import { BrowserRouter } from 'react-router-dom';
import List from '../../components/Applications/List';
import { applicationsStore } from '../../stores/Stores';

describe('Application List', () => {
  const minProps = {
    'applications': [{
      'id': '123',
      'name': 'ABC',
      'channels': [{
        'id': 'DEF',
        'name': 'main',
        'color': '#777777',
        'created_ts': '2018-10-16T21:07:56.819939+05:30',
        'application_id': '123',
        'package_id': 'XYZ',
        'package': {'id': 'PACK_ID', 'type': 4,
                    'version': '1.11.3', 'url': 'https://github.com/kinvolk',
                    'filename': '', 'description': '', 'size': '', 'hash': '', 'created_ts': '2019-07-18T20:10:39.163326+05:30',
                    'channels_blacklist': null, 'application_id': 'df1c8bbb-f525-49ee-8c94-3ca548b42059', 'flatcar_action': null, 'arch': 0},
        'arch': 0}],
      'description': 'App Item Description',
      'instances': {
        'count': '1'
      },
    }]

  };
  it('should render list correctly with data', () => {
    const getCachedApplications = applicationsStore.getCachedApplications();
    applicationsStore.getCachedApplications = () => minProps.applications;
    const {asFragment} = render(<BrowserRouter><List/></BrowserRouter>);
    expect(asFragment()).toMatchSnapshot();
    applicationsStore.getCachedApplications = getCachedApplications;
  });
});
