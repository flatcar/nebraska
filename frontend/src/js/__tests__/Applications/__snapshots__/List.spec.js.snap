// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Application List should render list correctly with data 1`] = `
<DocumentFragment>
  <div
    class="MuiGrid-root makeStyles-sectionHeader-2 MuiGrid-container MuiGrid-align-items-xs-flex-start MuiGrid-justify-xs-space-between"
  >
    <div
      class="MuiGrid-root MuiGrid-item"
    >
      <div
        class="MuiBox-root MuiBox-root-107"
      >
        <h4
          class="MuiTypography-root MuiTypography-h4"
        >
          Applications
        </h4>
      </div>
    </div>
    <div
      class="MuiGrid-root MuiGrid-item"
    >
      <div>
        <button
          aria-label="add"
          class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeSmall"
          data-testid="modal-button"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiIconButton-label"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root makeStyles-root-138 MuiSvgIcon-fontSizeLarge"
              focusable="false"
              role="presentation"
              viewBox="0 0 24 24"
            >
              <path
                d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
              />
            </svg>
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
  </div>
  <div
    class="MuiPaper-root MuiPaper-elevation1 MuiPaper-rounded"
  >
    <ul
      class="MuiList-root List-root-1 MuiList-padding"
    >
      <hr
        class="MuiDivider-root makeStyles-outterDivider-209"
        data-testid="list-item-divider"
      />
      <li
        class="MuiListItem-root makeStyles-root-207"
        data-testid="list-item"
      >
        <div
          class="MuiGrid-root MuiGrid-container"
        >
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
          >
            <div
              class="MuiGrid-root makeStyles-gridHeader-228 MuiGrid-container MuiGrid-justify-xs-space-between"
            >
              <div
                class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-1 MuiGrid-item MuiGrid-align-items-xs-center MuiGrid-justify-xs-space-between"
              >
                <div
                  class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
                >
                  <a
                    class="MuiTypography-root MuiLink-root MuiLink-underlineHover makeStyles-mainLink-230 MuiTypography-colorPrimary"
                    href="/apps/123"
                  >
                    ABC
                  </a>
                </div>
                <div
                  class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
                >
                  <div
                    class="MuiBox-root MuiBox-root-243"
                  >
                    <p
                      arial-label="group-id"
                      class="MuiTypography-root makeStyles-idLabel-233 MuiTypography-body1 MuiTypography-noWrap"
                    >
                      123
                    </p>
                  </div>
                  <div
                    class="MuiBox-root MuiBox-root-245"
                  >
                    <span
                      class="MuiTypography-root makeStyles-descriptionLabel-232 MuiTypography-body1"
                    >
                      App Item Description
                    </span>
                  </div>
                </div>
              </div>
              <div
                class="MuiGrid-root MuiGrid-item"
              >
                <div>
                  <button
                    aria-controls="simple-menu"
                    aria-haspopup="true"
                    class="MuiButtonBase-root MuiIconButton-root MuiIconButton-edgeEnd"
                    data-testid="more-menu-open-button"
                    tabindex="0"
                    type="button"
                  >
                    <span
                      class="MuiIconButton-label"
                    >
                      <svg
                        aria-hidden="true"
                        class="MuiSvgIcon-root"
                        focusable="false"
                        role="presentation"
                        viewBox="0 0 24 24"
                      >
                        <path
                          d="M12 8c1.1 0 2-.9 2-2s-.9-2-2-2-2 .9-2 2 .9 2 2 2zm0 2c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2zm0 6c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2z"
                        />
                      </svg>
                    </span>
                    <span
                      class="MuiTouchRipple-root"
                    />
                  </button>
                </div>
              </div>
            </div>
            <hr
              class="MuiDivider-root makeStyles-innerDivider-234 MuiDivider-light"
            />
          </div>
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
          >
            <div
              class="MuiGrid-root makeStyles-itemSection-208 MuiGrid-container"
            >
              <div
                class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-4"
              >
                <div
                  class="MuiBox-root MuiBox-root-250"
                >
                  <span
                    class="MuiTypography-root makeStyles-featureLabel-231 MuiTypography-body1"
                  >
                    INSTANCES
                  </span>
                  <span
                    class="MuiTypography-root makeStyles-label-235 makeStyles-label-252 MuiTypography-body1"
                  >
                    <h5
                      class="MuiTypography-root MuiTypography-h5"
                    >
                      1
                    </h5>
                    <div
                      class="MuiBox-root MuiBox-root-253"
                    >
                      <svg
                        aria-hidden="true"
                        class="MuiSvgIcon-root MuiSvgIcon-colorDisabled"
                        focusable="false"
                        role="presentation"
                        viewBox="0 0 24 24"
                      >
                        <path
                          d="M11.99 2C6.47 2 2 6.48 2 12s4.47 10 9.99 10C17.52 22 22 17.52 22 12S17.52 2 11.99 2zM12 20c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z"
                        />
                        <path
                          d="M12.5 7H11v6l5.25 3.15.75-1.23-4.5-2.67z"
                        />
                      </svg>
                      <div
                        class="MuiBox-root MuiBox-root-254"
                      >
                        <h6
                          class="MuiTypography-root MuiTypography-subtitle1"
                        >
                          last 24 hours
                        </h6>
                      </div>
                    </div>
                  </span>
                </div>
              </div>
              <div
                class="MuiBox-root MuiBox-root-255"
              >
                <hr
                  class="MuiDivider-root MuiDivider-vertical"
                />
              </div>
              <div
                class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-7"
              >
                <div
                  class="MuiBox-root MuiBox-root-256"
                >
                  <span
                    class="MuiTypography-root makeStyles-featureLabel-231 MuiTypography-body1"
                  >
                    Groups
                  </span>
                  <div
                    class="MuiBox-root MuiBox-root-258"
                  >
                    <span
                      class="MuiTypography-root makeStyles-label-235 makeStyles-label-259 MuiTypography-body1"
                    >
                      None
                    </span>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </li>
    </ul>
  </div>
</DocumentFragment>
`;
